File: lib/node/alphaskia/src/AlphaSkiaData.ts
import { AlphaSkiaNative } from './AlphaSkiaNative';
import { AlphaSkiaDataHandle, loadAddon } from './addon';

/**
 * @internal
 */
export class AlphaSkiaData extends AlphaSkiaNative<AlphaSkiaDataHandle> {
    public constructor(handle: AlphaSkiaDataHandle) {
        super(handle, loadAddon().alphaskia_data_free);
    }

    public toArray(): ArrayBuffer {
        this.checkDisposed();
        return loadAddon().alphaskia_data_get_data(this.handle!);
    }
}

File: lib/dotnet/AlphaSkia/AlphaSkiaData.cs
using System.Runtime.InteropServices;

namespace AlphaSkia;

internal sealed class AlphaSkiaData : AlphaSkiaNative
{
    public AlphaSkiaData(IntPtr handle)
        : base(handle, NativeMethods.alphaskia_data_free)
    {
    }

    public AlphaSkiaData(byte[] data)
        : base(NativeMethods.alphaskia_data_new_copy(data, (ulong)data.LongLength), NativeMethods.alphaskia_data_free)
    {
    }

    public byte[] ToArray()
    {
        CheckDisposed();
        var dataLength = NativeMethods.alphaskia_data_get_length(Handle);
        var data = new byte[dataLength];
        var dataPtr = NativeMethods.alphaskia_data_get_data(Handle);
        Marshal.Copy(dataPtr, data, 0, data.Length);
        return data;
    }
}